src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisWorkerService.java

```diff
@@ -125,26 +145,75 @@ private static String sanitizeParam(String param) {
-                            });
-                })
-                .doOnError(throwable -> LOGGER.error(throwable.toString(), throwable))
-                .subscribe();
+    public Consumer<Message<String>> consumeRun() {
+        return message -> {
+
+            SecurityAnalysisResultContext resultContext = SecurityAnalysisResultContext.fromMessage(message, objectMapper);
+            runRequests.add(resultContext.getResultUuid());
+
+            run(resultContext.getRunContext(), resultContext.getResultUuid())
+                    .flatMap(result -> resultRepository.insert(resultContext.getResultUuid(), result)
+                            .then(resultRepository.insertStatus(resultContext.getResultUuid(), SecurityAnalysisStatus.COMPLETED.name()))
+                            .then(Mono.just(result)))
+                    .doOnSuccess(result -> {
+                        if (result != null) {  // result available
+                            resultPublisherService.publish(resultContext.getResultUuid(), resultContext.getRunContext().getReceiver());
+                            LOGGER.info("Security analysis complete (resultUuid='{}')", resultContext.getResultUuid());
+                        } else {  // result not available : stop computation request
+                            if (cancelComputationRequests.get(resultContext.getResultUuid()) != null) {
+                                stoppedPublisherService.publish(resultContext.getResultUuid(), cancelComputationRequests.get(resultContext.getResultUuid()).getReceiver());
+                            }
+                        }
+                    })
+                    .doOnError(throwable -> {
+                        if (!(throwable instanceof CancellationException)) {
+                            LOGGER.error(throwable.toString(), throwable);
+                        }
+                    })
+                    .doFinally(s -> {
+                        futures.remove(resultContext.getResultUuid());
+                        cancelComputationRequests.remove(resultContext.getResultUuid());
+                        runRequests.remove(resultContext.getResultUuid());
+                    })
+                    .subscribe();
+        };
+    }
+
+    @Bean
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> {
+            SecurityAnalysisCancelContext cancelContext = SecurityAnalysisCancelContext.fromMessage(message);
+
+            if (runRequests.contains(cancelContext.getResultUuid())) {
+                cancelComputationRequests.put(cancelContext.getResultUuid(), cancelContext);
+            }
+
+            // find the completableFuture associated with result uuid
+            CompletableFuture<SecurityAnalysisResult> future = futures.get(cancelContext.getResultUuid());
+            if (future != null) {
+                future.cancel(true);  // cancel computation in progress
```

Comment:
do we want to cancel loading the networks and the contingencies ? It's an order of magnitude faster than doing the security analysis for a big analysis, but still it can take some time, I guess in the order of 1-10 seconds ?

Reply:
Maybe in another US ?